src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,394 @@
+  arma::Cube<ElemType> attnOut;
+
+  //! Locally-stored weight matrix associated with query.
+  OutputDataType queryWt;
+
+  //! Locally-stored weight matrix associated with key.
+  OutputDataType keyWt;
+
+  //! Locally-stored weight matrix associated with value.
+  OutputDataType valueWt;
+
+  //! Locally-stored weight matrix associated with attnWt.
+  OutputDataType outWt;
+
+  //! Locally-stored bias associated with query.
+  OutputDataType queryBias;
+
+  //! Locally-stored bias associated with key.
+  OutputDataType keyBias;
+
+  //! Locall-stored bias associated with value.
+  OutputDataType valueBias;
+
+  //! Locally-stored bias associated with attnWt.
+  OutputDataType outBias;
+
+  //! Locally-stored weights parameter.
+  OutputDataType weights;
+
+  //! Locally-stored projected query matrix over linear layer.
+  arma::Cube<ElemType> queryProjected;
+
+  //! Locally-stored projected key matrix over linear layer.
+  arma::Cube<ElemType> keyProjected;
+
+  //! Locally-stored projected value matrix over linear layer.
+  arma::Cube<ElemType> valueProjected;
+
+  //! Softmax layer to represent the probabilities of next sequence.
+  Softmax<InputDataType, OutputDataType> softmaxModule;
+
+  //! Dropout layer (optional).
+  Dropout<InputDataType, OutputDataType> dropoutModule;
+
+  //! Locally-stored delta object.
+  OutputDataType delta;
+
+  //! Locally-stored gradient object.
+  OutputDataType gradient;
+
+  //! Locally-stored output parameter.
+  OutputDataType outputParameter;
+}; // class MultiheadAttention
+
+/**
+ * Self Attention Policy class. It deals with the constraints of the parameters
+ * of the MultiheadAttention class.
+ */
+class SelfAttention
+{
+ public:
+
+  //! True if query, key and value are equal.
+  static const bool allInputsEqual = true;
+
+  //! True if key and value are equal.
+  static const bool keyEqualsValue = true;
+
+  /**
+   * Checks the validity of the dimensions of the input.
+   */
+  template <typename InputType>
+  static void CheckInputDimensions(const InputType& input,
+                                   const size_t tgtSeqLen,
+                                   const size_t srcSeqLen,
+                                   const size_t embedDim)
+  {
+    Log::Assert(tgtSeqLen == srcSeqLen);
+    Log::Assert(input.n_rows == tgtSeqLen * embedDim);
+  }
+
+  template <typename eT>
+  static void SplitInput(const arma::Mat<eT>& input,
+                         arma::Cube<eT>& query,
+                         arma::Cube<eT>& key,
+                         arma::Cube<eT>& value,
+                         const size_t tgtSeqLen,
+                         const size_t srcSeqLen,
+                         const size_t embedDim)
+  {
+    CheckInputDimensions(input, tgtSeqLen, srcSeqLen, embedDim);
+    query = arma::Cube<eT>(const_cast<arma::Mat<eT>&>(input).memptr(),
+        embedDim, tgtSeqLen, input.n_cols, false);
+    key = value = query;
```

Comment:
Do you need to copy the data here? Perhaps it's faster to use the same memptr constructor.

Reply:
I will better try pass three matrices concatenated into a single matrix as input as you suggested above.